bootstrap/controllers/machine_deployment_controller.go

```diff
@@ -0,0 +1,437 @@
+	// 1. It was not really understandable and mantainable.
+	// 2. Sometimes the reconciliation was not getting triggered when it should have, debugging this
+	// through the predicates was a nightmare.
+	if err := ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.MachineDeployment{}).
+		Owns(&clusterv1.Machine{}).
+		Complete(r); err != nil {
+		return fmt.Errorf("failed to get new controller builder: %w", err)
+	}
+
+	return nil
+}
+
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters;clusters/status,verbs=get;list;watch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machines;machines/status,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machinesets;machinesets/status,verbs=get;list;watch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machinedeployments;machinedeployments/status,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile handles the reconciliation of a MachineDeployment object.
+func (r *MachineDeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// NOTE(Hue): Generate a tarce ID for this reconciliation to track logs better.
+	traceID := trace.NewID()
+	log := r.Log.WithValues("machine_deployment", req.NamespacedName, "trace_id", traceID)
+	log.Info("Reconciliation started...")
+
+	machineDeployment := &clusterv1.MachineDeployment{}
+	if err := r.Get(ctx, req.NamespacedName, machineDeployment); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("MachineDeployment resource not found. Ignoring since object must be deleted.")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get MachineDeployment: %w", err)
+	}
+
+	if !machineDeployment.DeletionTimestamp.IsZero() {
+		log.Info("MachineDeployment is being deleted, skipping reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	ownedMachines, err := r.getOwnedMachines(ctx, machineDeployment)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get owned machines: %w", err)
+	}
+
+	scope, err := r.createScope(machineDeployment, ownedMachines)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create scope: %w", err)
+	}
+
+	if !r.hasUpgradeInstructions(scope) {
+		log.Info("MachineDeployment has no upgrade instructions, removing upgrade-to annotation from machines and stopping reconciliation")
+		if err := r.removeUpgradeToFromMachines(ctx, log, scope); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to remove upgrade-to annotation from machines: %w", err)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Starting the upgrade process
+	var upgradedMachines int
+	for _, m := range ownedMachines {
+		if r.isMachineUpgraded(scope, m) {
+			log.Info("Machine is already upgraded", "machine", m.Name)
+			upgradedMachines++
+			continue
+		}
+
+		if !m.DeletionTimestamp.IsZero() {
+			if !m.DeletionTimestamp.IsZero() {
```

Comment:
it is always good to double check but executing the same if statement twice in a nested block is probably not necessary ;)

Reply:
You know I can't stop being suuuper careful :smile:.
JK. Thanks for the catch!